File: py/structure/linear/Array.py
#!/usr/bin/env python
# -*- coding:gbk -*-

import sys

class Array:
    def __init__(self):
        self.__values, self.__length, self.__capacity = [0 for _ in range(64)], 0, 64

    def isEmpty(self) -> bool:
        return not self.__length

    def add(self, index: int, value: int) -> None:
        if index < 0 or index > self.__length:
            return

        if self.__length >= self.__capacity:
            self.__values.extend([0 for _ in range(16)])
            self.__capacity += 16

        for i in range(self.__length - 1, index - 1, -1):
            self.__values[i + 1] = self.__values[i]
        self.__values[index] = value
        self.__length += 1

    def addHead(self, value: int) -> None:
        if self.__length >= self.__capacity:
            self.__values.extend([0 for _ in range(16)])
            self.__capacity += 16

        for i in range(self.__length - 1, -1, -1):
            self.__values[i + 1] = self.__values[i]
        self.__values[0] = value
        self.__length += 1

    def addTail(self, value: int) -> None:
        if self.__length >= self.__capacity:
            self.__values.extend([0 for _ in range(16)])
            self.__capacity += 16

        self.__values[self.__length] = value
        self.__length += 1

    def erase(self, index: int) -> None:
        if index < 0 or index >= self.__length:
            return

        for i in range(index + 1, self.__length):
            self.__values[i - 1] = self.__values[i]
        self.__length -= 1

    def eraseHead(self, ) -> None:
        if not self.__length:
            return

        for i in range(1, self.__length):
            self.__values[i - 1] = self.__values[i]
        self.__length -= 1

    def eraseTail(self, ) -> None:
        if self.__length:
            self.__length -= 1

    def set(self, index: int, value: int) -> None:
        if index >= 0 and index < self.__length:
            self.__values[index] = value

    def get(self, index: int) -> int:
        if index >= 0 and index < self.__length:
            return self.__values[index]
        return -sys.maxsize

    def reverse(self) -> None:
        start, end = 0, self.__length - 1
        while start < end:
            self.__values[start], self.__values[end] = self.__values[end], self.__values[start]

            start += 1
            end -= 1

    def clear(self) -> None:
        self.__values, self.__length = [0 for _ in range(self.__capacity)], 0

File: cpp/structure/linear/Array.cpp
#include <iostream>
#include <vector>
using namespace std;

class Array {
public:
    Array() : values(vector<int>(64, 0)), length(0), capacity(64) {}

    bool isEmpty() {
        return !length;
    }

    void add(int index, int value) {
        if (index < 0 || index > length) {
            return;
        }

        if (length >= capacity) {
            values.resize(capacity + 16, 0);
            capacity += 16;
        }

        for (int i = length - 1; i >= index; --i) {
            values[i + 1] = values[i];
        }
        values[index] = value;
        ++length;
    }

    void addHead(int value) {
        if (length >= capacity) {
            values.resize(capacity + 16, 0);
            capacity += 16;
        }

        for (int i = length - 1; i >= 0; --i) {
            values[i + 1] = values[i];
        }
        values[0] = value;
        ++length;
    }

    void addTail(int value) {
        if (length >= capacity) {
            values.resize(capacity + 16, 0);
            capacity += 16;
        }

        values[length] = value;
        ++length;
    }

    void erase(int index) {
        if (index < 0 || index >= length) {
            return;
        }

        for (int i = index + 1; i < length; ++i) {
            values[i - 1] = values[i];
        }
        --length;
    }

    void eraseHead() {
        if (!length) {
            return;
        }

        for (int i = 1; i < length; ++i) {
            values[i - 1] = values[i];
        }
        --length;
    }

    void eraseTail() {
        if (length) {
            --length;
        }
    }

    void set(int index, int value) {
        if (index >= 0 && index < length) {
            values[index] = value;
        }
    }

    int get(int index) {
        if (index >= 0 && index < length) {
            return values[index];
        }
        return INT_MIN;
    }

    void reverse() {
        int start = 0;
        int end = length - 1;
        while (start < end) {
            swap(values[start], values[end]);

            ++start;
            --end;
        }
    }

    void clear() {
        values.assign(capacity, 0);
        length = 0;
    }

private:
    vector<int> values;
    int length;
    int capacity;
};
